### PR TITLE
Generate padding point using ChaCha20Rng directly

### PR DESCRIPTION
**Pull request by davxy**
Directly generate padding point using blake2b and ChaCha20Rng

### CODE REVIEW

File: ring/src/piop/params.rs [hunk 1]
```diff
@@ -32,7 +32,13 @@ pub struct PiopParams<F: PrimeField, Curve: SWCurveConfig<BaseField=F>> {
 
 impl<F: PrimeField, Curve: SWCurveConfig<BaseField=F>> PiopParams<F, Curve> {
     pub fn setup(domain: Domain<F>, h: Affine<Curve>, seed: Affine<Curve>) -> Self {
-        let padding_point = crate::hash_to_curve::<Affine<Curve>>(b"w3f/ring-proof/common/padding");
+        let padding_point = {
+            use ark_std::{rand::SeedableRng, UniformRand};
+            use blake2::Digest;
+            let seed = blake2::Blake2s::digest(b"w3f/ring-proof/common/padding");
```
Comment (davxy) [inline ring/src/piop/params.rs:38]: Do we want to use a more "neutral" seed here?

File: ring/src/piop/params.rs [hunk 2]
```diff
@@ -32,7 +32,13 @@ pub struct PiopParams<F: PrimeField, Curve: SWCurveConfig<BaseField=F>> {
 
 impl<F: PrimeField, Curve: SWCurveConfig<BaseField=F>> PiopParams<F, Curve> {
     pub fn setup(domain: Domain<F>, h: Affine<Curve>, seed: Affine<Curve>) -> Self {
-        let padding_point = crate::hash_to_curve::<Affine<Curve>>(b"w3f/ring-proof/common/padding");
+        let padding_point = {
+            use ark_std::{rand::SeedableRng, UniformRand};
+            use blake2::Digest;
+            let seed = blake2::Blake2b::digest(b"w3f/ring-proof/common/padding");
```
Comment (davxy) [inline ring/src/piop/params.rs:38]: Do we want to use a more "neutral" seed here?

Comment (burdges) [reply ring/src/piop/params.rs:38]: What do you mean?  An empty string or something?  Nah it's nice the string says something.

File: ring/src/piop/params.rs [hunk 3]
```diff
@@ -32,7 +32,13 @@ pub struct PiopParams<F: PrimeField, Curve: SWCurveConfig<BaseField=F>> {
...
+            Affine::<Curve>::rand(&mut rand_chacha::ChaCha20Rng::from_seed(seed.into()))
```
Comment (burdges) [inline ring/src/piop/params.rs:39]: This is a slightly slower try-and-increment hash to curve, which is fine, but just fyi.  The advantage of try-and-increment is that its universal.  It's not that slow either. 

https://github.com/search?q=repo%3Aarkworks-rs%2Falgebra%20path%3A%2F%5Eec%5C%2Fsrc%5C%2Fmodels%5C%2F%2F%20Distribution&type=code

https://github.com/arkworks-rs/algebra/blob/b33df5cce2d54cf4c9248e4b229c7d6708fa9375/ec/src/models/twisted_edwards/affine.rs#L97
https://github.com/arkworks-rs/algebra/blob/b33df5cce2d54cf4c9248e4b229c7d6708fa9375/ec/src/models/short_weierstrass/affine.rs#L105

**Review by burdges — APPROVED** [review]
LGTM but..

@swasilyev should confirm if this point needs to be in the prime order group.  Or if he needs a point in the coset somewhere.

File: ring/src/piop/params.rs (hunk 2 above)
Comment (davxy) [reply ring/src/piop/params.rs:38]: No I mean something not associated to any organization. Like "ring-proof-padding"

File: ring/src/piop/params.rs (hunk 3 above)
Comment (davxy) [reply ring/src/piop/params.rs:39]: Indeed I'd prefer a simpler try and increment. As it is easily replicable in a python script to store within the spec. 

We first need go generate the X coord (I'm assuming weierstrass) and then we recover Y using https://github.com/arkworks-rs/algebra/blob/b33df5cce2d54cf4c9248e4b229c7d6708fa9375/ec/src/models/short_weierstrass/affine.rs#L105

To generate X we use a hashing algorithm applied to something like `b"ring-proof-padding" || counter`. 
(counter is incremented in a loop until we don't find a valid point, which is expected to happen on average after 2 iterations). This is quite similar to the TAI method described in section 5.4.1.1 of https://datatracker.ietf.org/doc/rfc9381/

The only issue is, what hashing we use? I mean, if we use blake2 we can generate an X with at most 512 bits. Is this acceptable (I guess yes, at least yes for all the curves provided by arkworks)

**Comment by davxy** [comment]
Superseded by https://github.com/w3f/ring-proof/pull/32.

File: ring/src/piop/params.rs (hunk 3 above)
Comment (davxy) [reply ring/src/piop/params.rs:39]: https://github.com/w3f/ring-proof/pull/32

**Pull request closed without merging**